proposals/306-ipv6-happy-eyeballs.txt

```diff
@@ -48,135 +48,103 @@ Ticket: https://trac.torproject.org/projects/tor/ticket/29801
-   preferred address.
+   Bridges are out of scope for this proposal, because Tor does not support
+   multiple IP addresses in a single bridge line.
 
-   We should also allow falling back to the alternate address. For this,
-   three alternate designs will be given.
-
-3.1. Proposed Designs
-
-   This subsection will have three proposed designs for connecting to relays
-   via IPv4 and IPv6 in a Tor implementation of Happy Eyeballs.
-
-   These proposed designs will have some tradeoffs, including:
-
-    * Launching multiple TCP connections places up to 2x extra socket load on
-      dual-stack relays and authorities, because both connections may succeed.
-
-    * Launching multiple TLS connections places up to 2x the CPU load on
-      dual-stack relays and authorities, because both connections may succeed.
-
-    * Increasing the delay between connections mitigates the above issues,
-      but reduces perceived performance, particularly at bootstrap time
-      (pre-emptive circuits hide these delays after bootstrap).
+   The extend info structure should contain both an IPv4 and an IPv6 address.
+   This will allow us to try IPv4 and the IPv6 addresses should both be
+   available on a relay and the client is dual-stack.
 
-   The proposed designs are as listed as follows:
+   When parsing relay descriptors and filling in the extend info data
+   structure, we need to fill in both the IPv4 and IPv6 address if both are
+   available. If only one family is available for a relay (IPv4 or IPv6), we
+   should fill in the address for preferred family and leave the alternate
+   family null.
 
-    * Section 3.1.1: First Successful Authentication
+2.2. Guard Selection Changes
 
-    * Section 3.1.2: TCP Connection to Preferred Address On First Authenticated
-      Connection
+   In Proposal 283, we have allowed microdescriptor consensus documents
+   to contain IPv6 addresses. As most clients download microdescriptor
+   consensuses, Prop283 makes it possible to implement Prop306.
 
-    * Section 3.1.3: TCP Connection to Preferred Address On First TCP Success
+   When we select candidates for the entry node, we should have an option
+   for the number of entry guards. The parameter is as follows:
 
-3.1.1. First Successful Authentication
+     * NumIPv6Guards NUM
 
-   In this design, Tor will first connect to the preferred address and
-   attempt to authenticate. After a 1.5 second delay (based on Onionperf
-   data), Tor will connect to the alternate address and try to authenticate.
-   On the first successful authenticated connection, we close the other
-   connection.
+   If UseEntryGuards is set to 1, we will try to pick NUM routers that support
+   IPv6 for our primary guard list, which is the set of routers we strongly
+   prefer when connecting to the Tor network. If NUM is 0, we try to learn the
+   number from the guard-n-primary-guards-ipv6 consensus parameter, and default
+   to 1 if the consensus parameter isn’t set.
```

Comment:
The initial implementation should default to 0. Once we've tested it, we can increase the default to 1.
```suggestion
   to 0 if the consensus parameter isn’t set.
```

Reply:
I decided not to make this change.